test/test-close-fd.c

```diff
@@ -36,15 +37,36 @@ static void alloc_cb(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
 static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
   switch (++read_cb_called) {
   case 1:
-    ASSERT(nread == 1);
+    ASSERT_EQ(nread, 1);
     uv_read_stop(handle);
     break;
   case 2:
-    ASSERT(nread == UV_EOF);
+    ASSERT_EQ(nread, UV_EOF);
     uv_close((uv_handle_t *) handle, NULL);
     break;
   default:
-    ASSERT(!"read_cb_called > 2");
+    FATAL("read_cb_called > 2");
+  }
+}
+
+static void close_rx_read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
+  switch (++close_rx_read_cb_called) {
+  case 1:
+    ASSERT_EQ(nread, 1);
+    uv_read_stop(handle);
+    uv_close((uv_handle_t *) handle, NULL);
+    break;
+  case 2:
+#if defined(__FreeBSD__) || defined(__sun)
+    ASSERT_EQ(nread, UV_EOF);
+#else
+    ASSERT_EQ(nread, UV_EBADF);
+#endif
```

Comment:
@fantix I've been revisiting this. The notification of _the read-side pipe has been closed_ do vary depending on the OS. Is this what we really want?

Reply:
You made a good point.

> The notification of *the read-side pipe has been closed* do vary depending on the OS.

This is very true - on FreeBSD (and SmartOS too I think), `pipe()` yields 2 `O_RWDR` FDs, while Linux/macOS produces a pair of `O_RDONLY` + `O_WRONLY`. It makes more sense to listen on the write-side pipe and get an EOF as the notification of *the read-side pipe has been closed* in FreeBSD, rather than getting a `Bad file descriptor` for the same case on Linux/macOS. So maybe we could provide a consistent API with `UV_EOF`:

```diff
diff --git a/src/unix/stream.c b/src/unix/stream.c
index 0ecf46f5..81f8fe7b 100644
--- a/src/unix/stream.c
+++ b/src/unix/stream.c
@@ -1185,6 +1185,11 @@ static void uv__read(uv_stream_t* stream) {
       } else if (errno == ECONNRESET && stream->type == UV_NAMED_PIPE) {
         uv__stream_eof(stream, &buf);
         return;
 #endif
+#if !defined(_WIN32) && !defined(__FreeBSD__) && !defined(__sun)
+      } else if (errno == EBADF && stream->flags & UV_HANDLE_READABLE_CLOSED) {
+        uv__stream_eof(stream, &buf);
+        return;
+#endif
       } else {
         /* Error. User should call uv_close(). */
```

I will update this PR if this makes more sense?

Reply:
PR is now updated, please kindly review again. Much thanks!

Reply:
I'm not sure about this as I think this could potentially mask the other `EBADF` error cause  (for example some other thread could've closed the fd).

/cc @bnoordhuis @vtjnash do you have any thoughts on this, as you already commented on the original issue?

Reply:
That is the sort of event-system-implementation-dependent behavior I was concerned about in the original issue, though I don't have any particular insights to offer.